hugolib/page__paths.go

```diff
@@ -141,6 +141,21 @@ func createTargetPathDescriptor(p *pageState) (page.TargetPathDescriptor, error)
 	desc.PrefixFilePath = s.getLanguageTargetPathLang(alwaysInSubDir)
 	desc.PrefixLink = s.getLanguagePermalinkLang(alwaysInSubDir)
 
+	if desc.URL != "" && strings.IndexByte(desc.URL, ':') >= 0 {
+		// Attempt to parse and expand an url. Probably it's a good
+		// idea to cache the parsed url to avoid regexp matching and memory allocations.
+
+		opath, err := d.ResourceSpec.Permalinks.ExpandPattern(desc.URL, p)
```

Comment:
You are right about the "cache" remark.

1. The parser isn't fast.
2. I predict that 90% of the use of this new feature will be via cascade, so it should be cache friendly.

* You can get a thread safe cache with `maps.NewCache`
* I suggest you use the `GetOrCreate` method there; I also suggest you rename the `ExpandPattern` to `GetOrExpandPattern` to reflect this.